File: src/main/java/no/nav/pensjonbrevdata/model/DoksysbrevV2.java
package no.nav.pensjonbrevdata.model;

import kotlin.jvm.functions.Function1;
import no.nav.pensjonbrevdata.model.codes.*;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.function.Supplier;
import java.util.stream.Collectors;

public class DoksysbrevV2 extends Doksysbrev {
    public DoksysbrevV2(String brevkodeInBrevsystem, boolean redigerbart, String dekode, BrevkategoriCode brevkategori, DokumenttypeCode doktype, List<SprakCode> sprak, Boolean visIPselv, BrevUtlandCode utland, BrevregeltypeCode brevregeltype, BrevkravtypeCode brevkravtype, DokumentkategoriCode dokumentkategori, Boolean synligForVeileder, BrevkontekstCode brevkontekst, Integer prioritet, String dokumentmalId, String dokumentmalFelleselementId, Supplier<List<DoksysVedlegg>> vedleggListe) {
        super(brevkodeInBrevsystem, redigerbart, dekode, brevkategori, doktype, sprak, visIPselv, utland, brevregeltype, brevkravtype, dokumentkategori, synligForVeileder, brevkontekst, prioritet, dokumentmalId, dokumentmalFelleselementId, vedleggListe);
    }

    @Override
    public Brevdata medXSD(@NotNull Function1<String, String> dokumentmalGenerator, @NotNull Function1<String, String> fellesmalGenerator) {
        String dokumentmal = dokumentmalGenerator.invoke("v2." + dokumentmalId);
        String fellesmal = fellesmalGenerator.invoke(dokumentmalFelleselementId);
        Supplier<List<DoksysVedlegg>> vedleggListeMedXSD = vedleggListe == null ? null : () -> vedleggListe.get().stream().map((vedlegg) -> vedlegg.medXSD(dokumentmalGenerator, fellesmalGenerator)).collect(Collectors.toList());
        return new Doksysbrev(getBrevkodeIBrevsystem(), isRedigerbart(), getDekode(), getBrevkategori(), getDokType(),
                getSprak(), getVisIPselv(), getUtland(), getBrevregeltype(), getBrevkravtype(), getDokumentkategori(),
                getSynligForVeileder(), getBrevkontekst(), getPrioritet(), vedleggListeMedXSD, dokumentmalId,
                dokumentmalFelleselementId, dokumentmal, fellesmal);
    }

}

File: src/main/java/no/nav/pensjonbrevdata/model/DoksysbrevV2.kt
package no.nav.pensjonbrevdata.model

import no.nav.pensjonbrevdata.model.codes.*
import java.util.function.Supplier

class DoksysbrevV2(
    brevkodeInBrevsystem: String,
    redigerbart: Boolean,
    dekode: String,
    brevkategori: BrevkategoriCode?,
    doktype: DokumenttypeCode,
    sprak: List<SprakCode>?,
    visIPselv: Boolean?,
    utland: BrevUtlandCode?,
    brevregeltype: BrevregeltypeCode?,
    brevkravtype: BrevkravtypeCode?,
    dokumentkategori: DokumentkategoriCode,
    synligForVeileder: Boolean?,
    brevkontekst: BrevkontekstCode?,
    prioritet: Int?,
    dokumentmalId: String,
    dokumentmalFelleselementId: String,
    vedleggListe: Supplier<List<DoksysVedlegg>>?
) : Doksysbrev(
    brevkodeIBrevsystem = brevkodeInBrevsystem,
    redigerbart = redigerbart,
    dekode = dekode,
    brevkategori = brevkategori,
    dokType = doktype,
    sprak = sprak,
    visIPselv = visIPselv,
    utland = utland,
    brevregeltype = brevregeltype,
    brevkravtype = brevkravtype,
    dokumentkategori = dokumentkategori,
    synligForVeileder = synligForVeileder,
    brevkontekst = brevkontekst,
    prioritet = prioritet,
    vedleggListe = vedleggListe,
    dokumentmalId = dokumentmalId,
    dokumentmalFelleselementId = dokumentmalFelleselementId,
) {
    override fun medXSD(
        dokumentmalGenerator: (String) -> String,
        fellesmalGenerator: (String) -> String
    ): Brevdata {
        val dokumentmal = dokumentmalGenerator("v2.$dokumentmalId")
        val fellesmal = fellesmalGenerator(dokumentmalFelleselementId)
        val vedleggListeMedXSD = vedleggListe?.let { liste ->
            Supplier {
                liste.get().map { vedlegg -> vedlegg.medXSD(dokumentmalGenerator, fellesmalGenerator) }
            }
        }
        return Doksysbrev(
            brevkodeIBrevsystem, redigerbart, dekode, brevkategori, dokType,
            sprak, visIPselv, utland, brevregeltype, brevkravtype, dokumentkategori,
            synligForVeileder, brevkontekst, prioritet,  dokumentmalId,
            dokumentmalFelleselementId, vedleggListeMedXSD, dokumentmal, fellesmal
        )
    }
}
